### PR TITLE
*: Driver install and remove handling

### PR DESCRIPTION
**Pull request by mcuadros**
New methods for bblfshctl and daemon, also all the packages has been updated and reviewed to being able to handle properly this actions.

### CODE REVIEW

**Comment by mcuadros** [comment]
Fixes https://github.com/bblfsh/server/issues/80
Fixes https://github.com/bblfsh/server/issues/32
Fixes https://github.com/bblfsh/server/issues/28

File: daemon/protocol/service.go [hunk 1]
```diff
@@ -81,3 +83,60 @@ func DriverStates() *DriverStatesResponse {
...
+
+//proteus:generate
+type InstallDriverRequest struct {
+	// Language supported by the driver being installed.
+	Language string
+	// ImageReference name of the image to be installed in the following format:
+	// `[transport://]name[:tag]` if transport is missing `docker` or tag is
+	// missing `lastest` are assumed.
+	ImageReference string
+	// Update if update if false install fails if an previous image already
```
Comment (abeaumont) [inline daemon/protocol/service.go]: `// Update indicates whether an image should be updated. When set to false, the installation fails if the image already exists.`

File: daemon/protocol/service.go [hunk 2]
```diff
@@ -81,3 +83,60 @@ func DriverStates() *DriverStatesResponse {
 
 	return resp
 }
+
+//proteus:generate
+type InstallDriverRequest struct {
+	// Language supported by the driver being installed.
+	Language string
+	// ImageReference name of the image to be installed in the following format:
+	// `[transport://]name[:tag]` if transport is missing `docker` or tag is
```
Comment (abeaumont) [inline daemon/protocol/service.go]: * s/ImageReference name/ImageReference is the name/
* s/if transport is missing `docker` or tag is missing `lastest` are assumed./Default values are: `docker` for transport and `latest` for tag/

File: cli/bblfshctl/cmd/driver.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+const (
+	DriverCommandDescription = "Manage drivers: install, remove and list"
+	DriverCommandHelp        = DriverCommandDescription
+)
+
+type DriverCommand struct {
+	ControlCommand
+}
+
+func (*DriverCommand) Execute([]string) error {
```
Comment (abeaumont) [inline cli/bblfshctl/cmd/driver.go]: Is this needed?

File: cli/bblfshctl/cmd/parse.go [hunk 4]
```diff
@@ -81,13 +85,18 @@ func (c *ParseCommand) doNative(content string) error {
 func printResponse(r *protocol.Response) {
 	fmt.Printf("Status: %s\n", r.Status)
 	fmt.Printf("Elapsed: %s\n", r.Elapsed)
+	printErrors(r.Errors)
+	printErrors(r.Errors)
```
Comment (abeaumont) [inline cli/bblfshctl/cmd/parse.go]: Duplicated

File: cli/bblfshctl/cmd/parse.go [hunk 5]
```diff
@@ -81,13 +85,18 @@ func (c *ParseCommand) doNative(content string) error {
...
+	if len(errors) != 0 {
 		fmt.Println("Errors:")
-		for _, err := range r.Errors {
+		for _, err := range errors {
 			fmt.Printf("\t- %s\n", err)
 		}
-	}
 
-	fmt.Println("")
+		os.Exit(1)
```
Comment (abeaumont) [inline cli/bblfshctl/cmd/parse.go]: I don't think `printErrors` should finish the program.

File: cli/bblfshctl/main.go [hunk 6]
```diff
@@ -16,17 +16,49 @@ var (
...
 
 	if _, err := parser.Parse(); err != nil {
-		if _, ok := err.(*flags.Error); ok {
+		if flagsErr, ok := err.(*flags.Error); ok && flagsErr.Type == flags.ErrHelp {
+			os.Exit(0)
+		} else {
+			fmt.Println()
 			parser.WriteHelp(os.Stdout)
-			fmt.Printf("\nBuild information\n  commit: %s\n  date: %s\n", version, build)
+			fmt.Printf("\nBuild information\n  commit: %s\n  date:%s\n", version, build)
```
Comment (abeaumont) [inline cli/bblfshctl/main.go]: Why is this line changed?

File: daemon/protocol/types.go [hunk 7]
```diff
@@ -67,13 +66,9 @@ type DriverImageState struct {
...
 	Build time.Time `json:"build,omitempty"`
-	// Status is the development status of the driver (alpha, beta, etc).
-	Status string `json:"status"`
-	// OS is the linux distribution running on the driver container.
-	OS string `json:"os"`
-	// Native version is the version of the compiler/interpreter being use in the
-	// native side of the driver.
+	// Status is the development status of the driver (alpha, beta, etc)
+	Status        string   `json:"status"`
+	OS            string   `json:"os"`
```
Comment (abeaumont) [inline daemon/protocol/types.go]: Why is documentation removed?

File: runtime/container_test.go [hunk 8]
```diff
@@ -23,7 +23,7 @@ func TestContainerSuite(t *testing.T) {
 }
 
 func (s *ContainerSuite) SetupSuite() {
-	IfNetworking(s.T())
+	//	IfNetworking(s.T())
```
Comment (abeaumont) [inline runtime/container_test.go]: leftover?

File: cli/bblfshctl/cmd/driver.go (hunk 3 above)
Comment (mcuadros) [reply cli/bblfshctl/cmd/driver.go]: it should fit the "interface"

File: cli/bblfshctl/main.go (hunk 6 above)
Comment (mcuadros) [reply cli/bblfshctl/main.go]: the missing space

Comment (abeaumont) [reply cli/bblfshctl/main.go]: I see there's now a missing space, I don't see the point of removing that space, that's what I was asking.

File: cli/bblfshctl/cmd/driver.go (hunk 3 above)
Comment (abeaumont) [reply cli/bblfshctl/cmd/driver.go]: This is clearly an _abstract_ struct/class/whatever. Is still needed to implement a dummy method? If so, just ignore this comment.

File: cli/bblfshctl/main.go (hunk 6 above)
Comment (mcuadros) [reply cli/bblfshctl/main.go]: because you have a space on commit, so you should have one in the date too. if not looks wrong

File: cli/bblfshctl/cmd/driver.go (hunk 3 above)
Comment (mcuadros) [reply cli/bblfshctl/cmd/driver.go]: Is made by reflection, is search for a Execute method, otherwise will panic in runtime.

File: cli/bblfshctl/main.go (hunk 6 above)
Comment (abeaumont) [reply cli/bblfshctl/main.go]: You're removing the space, not adding it.
